### PR TITLE
update error handling in claim based access validation

### PR DESCRIPTION
**Pull request by BiyonFernando**

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/mediators/ClaimBasedResourceAccessValidationMediator.java [hunk 1]
```diff
@@ -62,42 +62,45 @@ public boolean mediate(MessageContext messageContext) {
...
-                handleFailure(HttpStatus.SC_FORBIDDEN, messageContext, CLAIMS_MISMATCH_ERROR_MSG, null);
-                return false;
+                log.debug("A regex is not provided, validating the claim values based on equality.");
+                if ((StringUtils.equals(accessVerificationClaimValue, claimValueSentInToken)) || shouldAllowValidation) {
+                    log.debug("Claim values match or the flow is configured to allow when claims doesn't match. " +
+                            "Hence the flow is allowed.");
+                    return true;
+                } else {
+                    log.debug("Claim values don't match. Hence the flow is not allowed.");
+                    throw new APISecurityException(APISecurityConstants.API_AUTH_ACCESS_TOKEN_CLAIMS_MISMATCH,APISecurityConstants.API_AUTH_ACCESS_TOKEN_CLAIMS_MISMATCH_MESSAGE, APISecurityConstants.API_AUTH_ACCESS_TOKEN_CLAIMS_MISMATCH_DESCRIPTION);
```
Comment (PasanT9) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/mediators/ClaimBasedResourceAccessValidationMediator.java:96]: This line exceeds 120 chars. Shall we fix it? Please check other places as well.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/mediators/ClaimBasedResourceAccessValidationMediator.java [hunk 2]
```diff
@@ -62,42 +62,45 @@ public boolean mediate(MessageContext messageContext) {
...
+                        "claim \"%s\"", accessVerificationClaim));
             }
-        } else {
-            log.debug("A regex is not provided, validating the claim values based on equality.");
-            if ((StringUtils.equals(accessVerificationClaimValue, claimValueSentInToken)) || shouldAllowValidation) {
-                log.debug("Claim values match or the flow is configured to allow when claims doesn't match. " +
-                        "Hence the flow is allowed.");
-                return true;
+
+            if (StringUtils.isNotBlank(accessVerificationClaimValueRegex)) {
```
Comment (PasanT9) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/mediators/ClaimBasedResourceAccessValidationMediator.java:73]: Shall we replace the if and else blocks to improve readability

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/mediators/ClaimBasedResourceAccessValidationMediator.java [hunk 3]
```diff
@@ -62,42 +62,45 @@ public boolean mediate(MessageContext messageContext) {
...
+                Matcher tokenSentClaimValueMatcher = pattern.matcher(claimValueSentInToken);
+
+                if ((configuredClaimValueMatcher.matches() && tokenSentClaimValueMatcher.matches())
+                        || shouldAllowValidation) {
+                    log.debug("Claim values match or the flow is configured to allow when claims doesn't match. " +
+                            "Hence the flow is allowed.");
+                    return true;
+                } else {
+                    log.debug("Claim values don't match. Hence the flow is not allowed.");
+                    throw new APISecurityException(APISecurityConstants.API_AUTH_ACCESS_TOKEN_CLAIMS_MISMATCH,APISecurityConstants.API_AUTH_ACCESS_TOKEN_CLAIMS_MISMATCH_MESSAGE, APISecurityConstants.API_AUTH_ACCESS_TOKEN_CLAIMS_MISMATCH_DESCRIPTION);
```
[the remark this answers is not in the document]
Comment (BiyonFernando) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/mediators/ClaimBasedResourceAccessValidationMediator.java]: Now it is formatted correctly

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/mediators/ClaimBasedResourceAccessValidationMediator.java (hunk 1 above)
Comment (BiyonFernando) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/mediators/ClaimBasedResourceAccessValidationMediator.java]: Now it is formatted correctly

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/mediators/ClaimBasedResourceAccessValidationMediator.java (hunk 2 above)
Comment (BiyonFernando) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/mediators/ClaimBasedResourceAccessValidationMediator.java]: I put the else block to improve the readability.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/APISecurityConstants.java [hunk 4]
```diff
@@ -138,26 +154,32 @@ public static final String getAuthenticationFailureMessage(int errorCode) {
...
-            case API_AUTH_ACCESS_TOKEN_EXPIRED:
-                errorDescription += DESCRIPTION_SEPARATOR + API_AUTH_ACCESS_TOKEN_EXPIRED_DESCRIPTION;
-                break;
-            case API_AUTH_INVALID_CREDENTIALS:
-                errorDescription += DESCRIPTION_SEPARATOR + API_AUTH_INVALID_CREDENTIALS_DESCRIPTION;
-                break;
-            default:
-                // Do nothing since we are anyhow returning the original error description.
+        switch (errorCode) {
+        case API_AUTH_INCORRECT_API_RESOURCE:
```
Comment (PasanT9) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/APISecurityConstants.java:160]: Shall we fix the indentations in switch block?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/mediators/ClaimBasedResourceAccessValidationMediator.java [hunk 5]
```diff
@@ -62,41 +61,50 @@ public boolean mediate(MessageContext messageContext) {
...
-            log.debug("A regex is not provided, validating the claim values based on equality.");
-            if ((StringUtils.equals(accessVerificationClaimValue, claimValueSentInToken)) || shouldAllowValidation) {
-                log.debug("Claim values match or the flow is configured to allow when claims doesn't match. " +
-                        "Hence the flow is allowed.");
-                return true;
-            } else {
-                log.debug("Claim values don't match. Hence the flow is not allowed.");
-                handleFailure(HttpStatus.SC_FORBIDDEN, messageContext, CLAIMS_MISMATCH_ERROR_MSG, null);
-                return false;
+
```
Comment (PasanT9) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/mediators/ClaimBasedResourceAccessValidationMediator.java:72]: Shall we remove this empty line as well?

**Review by PasanT9 — APPROVED** [review]

**Pull request merged**